File: csharp/src/LabelSuite.Core/Schema.cs
// 검사 목록 표준 스키마 — 컬럼 계약의 단일 원천 (파이썬 core/schema.py 포팅).
using System.Text.RegularExpressions;
using ClosedXML.Excel;

namespace LabelSuite.Core;

public class SchemaException(string message) : Exception(message);

public sealed record LabelRecord(
    string Lot, string Products, string Pn, string Ref,
    string MfgDate, string ExpDate, string Gtin, string? Standard = null)
{
    public string Field(string column) => column switch
    {
        "LOT" => Lot, "PRODUCTS" => Products, "PN" => Pn, "REF" => Ref,
        "MFG DATE" => MfgDate, "EXP DATE" => ExpDate, "GTIN" => Gtin,
        _ => throw new ArgumentOutOfRangeException(nameof(column), column, null),
    };
}

public static class Schema
{
    public static readonly string[] CanonicalColumns =
        ["LOT", "PRODUCTS", "PN", "REF", "MFG DATE", "EXP DATE", "GTIN"];

    public const string StandardColumn = "STANDARD";
    public const string SheetName = "Label Inspection List";
    public const string DateFormatDefault = "yyyy-MM-dd";

    private static readonly Regex FloatArtifact = new(@"^\d+\.0+$", RegexOptions.Compiled);
    private static readonly Regex Scientific = new(@"^[\d.]+[eE]\+?\d+$", RegexOptions.Compiled);
    private static readonly Regex Gs1Paren = new(@"^\(01\)(\d{1,14})$", RegexOptions.Compiled);
    private static readonly Regex Gs1Bare = new(@"^01(\d{14})$", RegexOptions.Compiled);
    private static readonly Regex NonDigit = new(@"\D", RegexOptions.Compiled);

    /// <summary>GTIN 정규화의 유일한 규칙: 숫자만 남겨 14자리 zero-pad.</summary>
    public static string NormalizeGtin14(object? raw)
    {
        if (raw is null) return "";
        var text = Convert.ToString(raw, System.Globalization.CultureInfo.InvariantCulture)?.Trim() ?? "";
        if (text.Length == 0 || text.Equals("nan", StringComparison.OrdinalIgnoreCase)
            || text.Equals("none", StringComparison.OrdinalIgnoreCase)) return "";
        if (FloatArtifact.IsMatch(text)) text = text.Split('.')[0];
        else if (Scientific.IsMatch(text))
            text = ((long)double.Parse(text, System.Globalization.CultureInfo.InvariantCulture)).ToString();
        var m = Gs1Paren.Match(text);
        if (!m.Success) m = Gs1Bare.Match(text);
        if (m.Success) text = m.Groups[1].Value;
        var digits = NonDigit.Replace(text, "");
        if (digits.Length == 0) return "";
        if (digits.Length > 14) throw new SchemaException($"GTIN이 14자리를 초과합니다: {raw}");
        return digits.PadLeft(14, '0');
    }

    /// <summary>규격별 날짜 포맷 폴백 규칙 (standards.json의 date_format이 우선).</summary>
    public static string DateFormatFor(string? standard) =>
        standard == "BSC" ? "yyyy.MM.dd" : DateFormatDefault;

    private static string CleanCell(object? value)
    {
        var text = Convert.ToString(value, System.Globalization.CultureInfo.InvariantCulture)?.Trim() ?? "";
        return text.Equals("nan", StringComparison.OrdinalIgnoreCase)
            || text.Equals("none", StringComparison.OrdinalIgnoreCase) ? "" : text;
    }

    /// <summary>7컬럼(레거시) 또는 8컬럼 목록 파일 로드. GTIN은 GTIN-14로 정규화.</summary>
    public static (List<LabelRecord> Records, List<string> Warnings) LoadInspectionList(string path)
    {
        using var workbook = new XLWorkbook(path);
        var sheet = workbook.Worksheets.Contains(SheetName)
            ? workbook.Worksheet(SheetName) : workbook.Worksheets.First();
        var rows = sheet.RowsUsed().ToList();
        if (rows.Count == 0) throw new SchemaException("목록 파일이 비어 있습니다.");

        var header = rows[0].CellsUsed().Select(c => CleanCell(c.Value)).ToList();
        var missing = CanonicalColumns.Where(c => !header.Contains(c)).ToList();
        if (missing.Count > 0)
            throw new SchemaException(
                $"목록 파일의 컬럼이 올바르지 않습니다. 누락: {string.Join(", ", missing)}");
        var index = CanonicalColumns.ToDictionary(c => c, c => header.IndexOf(c) + 1);
        var hasStandard = header.Contains(StandardColumn);
        var standardIndex = hasStandard ? header.IndexOf(StandardColumn) + 1 : -1;

        var records = new List<LabelRecord>();
        var warnings = new List<string>();
        foreach (var row in rows.Skip(1))
        {
            string Cell(string name) => CleanCell(row.Cell(index[name]).GetFormattedString());
            var all = CanonicalColumns.Select(Cell).ToList();
            if (all.All(string.IsNullOrEmpty)) continue;
            string gtin;
            try { gtin = NormalizeGtin14(Cell("GTIN")); }
            catch (SchemaException ex)
            {
                warnings.Add($"{row.RowNumber()}행: {ex.Message} — 원본 값을 유지합니다.");
                gtin = Cell("GTIN");
            }
            var standard = hasStandard ? CleanCell(row.Cell(standardIndex).GetFormattedString()) : "";
            records.Add(new LabelRecord(
                Cell("LOT"), Cell("PRODUCTS"), Cell("PN"), Cell("REF"),
                Cell("MFG DATE"), Cell("EXP DATE"), gtin,
                standard.Length > 0 ? standard : null));
        }
        if (records.Count == 0) warnings.Add("목록에 데이터 행이 없습니다.");
        return (records, warnings);
    }

    /// <summary>목록 저장. GTIN 셀은 텍스트 서식으로 선행 0 보존.</summary>
    public static void SaveInspectionList(IEnumerable<LabelRecord> records, string path,
                                          bool includeStandard = true)
    {
        using var workbook = new XLWorkbook();
        var sheet = workbook.AddWorksheet(SheetName);
        var header = CanonicalColumns.ToList();
        if (includeStandard) header.Add(StandardColumn);
        for (var i = 0; i < header.Count; i++) sheet.Cell(1, i + 1).Value = header[i];

        var gtinColumn = Array.IndexOf(CanonicalColumns, "GTIN") + 1;
        var rowNumber = 2;
        foreach (var record in records)
        {
            var values = new List<string>
                { record.Lot, record.Products, record.Pn, record.Ref,
                  record.MfgDate, record.ExpDate, record.Gtin };
            if (includeStandard) values.Add(record.Standard ?? "");
            for (var i = 0; i < values.Count; i++)
            {
                var cell = sheet.Cell(rowNumber, i + 1);
                if (i + 1 == gtinColumn) cell.Style.NumberFormat.Format = "@";
                cell.SetValue(values[i]);
            }
            rowNumber++;
        }
        sheet.SheetView.FreezeRows(1);
        sheet.Columns().AdjustToContents();
        workbook.SaveAs(path);
    }
}

File: labelsuite/core/schema.py
"""검사 목록의 표준 스키마 — 두 레거시 프로그램에 중복돼 있던 컬럼 계약의 단일 원천.

목록 파일 계약:
- 시트명 'Label Inspection List'
- 필수 컬럼 7개(CANONICAL_COLUMNS) + 선택적 'STANDARD' 컬럼(검사 규격 자동 선택용)
- GTIN은 항상 GTIN-14(14자리, 선행 0 유지)로 저장
"""

from __future__ import annotations

import re
from dataclasses import dataclass, replace
from typing import Iterable

from openpyxl import Workbook, load_workbook

CANONICAL_COLUMNS = ("LOT", "PRODUCTS", "PN", "REF", "MFG DATE", "EXP DATE", "GTIN")
STANDARD_COLUMN = "STANDARD"
SHEET_NAME = "Label Inspection List"

# 날짜 표기 규격: BSC(일본)는 점 구분, 그 외는 하이픈 구분
DATE_FORMAT_DEFAULT = "%Y-%m-%d"
DATE_FORMAT_BY_STANDARD = {"BSC": "%Y.%m.%d"}


class SchemaError(ValueError):
    """목록 파일이 스키마 계약을 위반했을 때."""


@dataclass(frozen=True)
class LabelRecord:
    """검사 목록의 한 행."""

    lot: str
    products: str
    pn: str
    ref: str
    mfg_date: str
    exp_date: str
    gtin: str
    standard: str | None = None

    def as_row(self, include_standard: bool) -> list[str]:
        row = [self.lot, self.products, self.pn, self.ref,
               self.mfg_date, self.exp_date, self.gtin]
        if include_standard:
            row.append(self.standard or "")
        return row

    def field(self, column: str) -> str:
        return {
            "LOT": self.lot, "PRODUCTS": self.products, "PN": self.pn,
            "REF": self.ref, "MFG DATE": self.mfg_date,
            "EXP DATE": self.exp_date, "GTIN": self.gtin,
        }[column]


def normalize_gtin14(raw: object) -> str:
    """GTIN 정규화의 유일한 규칙: 숫자만 남겨 14자리로 zero-pad.

    엑셀에서 흘러들어오는 float 흔적('8806173612345.0'), int, 문자열을 모두 수용한다.
    빈 값은 빈 문자열 그대로, 14자리 초과는 SchemaError.
    """
    if raw is None:
        return ""
    text = str(raw).strip()
    if not text or text.lower() in ("nan", "none"):
        return ""
    # 엑셀 float 흔적 제거: '123.0' → '123' (지수 표기 포함)
    if re.fullmatch(r"\d+\.0+", text):
        text = text.split(".")[0]
    elif re.fullmatch(r"[\d.]+[eE]\+?\d+", text):
        text = format(int(float(text)), "d")
    # GS1 AI(01) 접두가 붙은 바코드 텍스트 수용: '(01)08806...' / '0108806...'(16자리)
    match = re.fullmatch(r"\(01\)(\d{1,14})", text) or re.fullmatch(r"01(\d{14})", text)
    if match:
        text = match.group(1)
    digits = re.sub(r"\D", "", text)
    if not digits:
        return ""
    if len(digits) > 14:
        raise SchemaError(f"GTIN이 14자리를 초과합니다: {raw!r}")
    return digits.zfill(14)


def date_format_for(standard: str | None) -> str:
    """규격별 날짜 포맷. standards.json의 date_format이 우선하며 이는 폴백 규칙이다."""
    if standard and standard in DATE_FORMAT_BY_STANDARD:
        return DATE_FORMAT_BY_STANDARD[standard]
    return DATE_FORMAT_DEFAULT


def _clean_cell(value: object) -> str:
    if value is None:
        return ""
    text = str(value).strip()
    return "" if text.lower() in ("nan", "none") else text


def load_inspection_list(path: str) -> tuple[list[LabelRecord], list[str]]:
    """7컬럼(레거시) 또는 8컬럼 목록 파일을 읽는다.

    반환: (records, warnings). 헤더가 계약과 다르면 SchemaError.
    레거시 파일의 13자리 GTIN은 로드 시 GTIN-14로 정규화된다.
    """
    wb = load_workbook(path, read_only=True, data_only=True)
    try:
        ws = wb[SHEET_NAME] if SHEET_NAME in wb.sheetnames else wb.worksheets[0]
        rows = ws.iter_rows(values_only=True)
        try:
            header = next(rows)
        except StopIteration:
            raise SchemaError("목록 파일이 비어 있습니다.")
        header = [_clean_cell(h) for h in header if h is not None]

        missing = [c for c in CANONICAL_COLUMNS if c not in header]
        if missing:
            raise SchemaError(
                "목록 파일의 컬럼이 올바르지 않습니다. "
                f"누락: {', '.join(missing)} / 발견된 헤더: {header}"
            )
        col_idx = {name: header.index(name) for name in CANONICAL_COLUMNS}
        has_standard = STANDARD_COLUMN in header
        if has_standard:
            col_idx[STANDARD_COLUMN] = header.index(STANDARD_COLUMN)

        records: list[LabelRecord] = []
        warnings: list[str] = []
        for i, row in enumerate(rows, start=2):
            if row is None or all(v is None or _clean_cell(v) == "" for v in row):
                continue

            def cell(name: str) -> str:
                idx = col_idx[name]
                return _clean_cell(row[idx]) if idx < len(row) else ""

            try:
                gtin = normalize_gtin14(cell("GTIN"))
            except SchemaError as exc:
                warnings.append(f"{i}행: {exc} — 원본 값을 유지합니다.")
                gtin = cell("GTIN")
            records.append(LabelRecord(
                lot=cell("LOT"), products=cell("PRODUCTS"), pn=cell("PN"),
                ref=cell("REF"), mfg_date=cell("MFG DATE"),
                exp_date=cell("EXP DATE"), gtin=gtin,
                standard=(cell(STANDARD_COLUMN) or None) if has_standard else None,
            ))
        if not records:
            warnings.append("목록에 데이터 행이 없습니다.")
        return records, warnings
    finally:
        wb.close()


def save_inspection_list(records: Iterable[LabelRecord], path: str,
                         include_standard: bool = True) -> None:
    """openpyxl로 저장. GTIN 셀은 텍스트 서식으로 선행 0을 보존한다."""
    wb = Workbook()
    ws = wb.active
    ws.title = SHEET_NAME
    header = list(CANONICAL_COLUMNS)
    if include_standard:
        header.append(STANDARD_COLUMN)
    ws.append(header)
    gtin_col = CANONICAL_COLUMNS.index("GTIN") + 1
    for record in records:
        ws.append(record.as_row(include_standard))
        ws.cell(row=ws.max_row, column=gtin_col).number_format = "@"
    ws.freeze_panes = "A2"
    for col_cells in ws.columns:
        width = max(len(str(c.value or "")) for c in col_cells)
        ws.column_dimensions[col_cells[0].column_letter].width = min(width + 2, 40)
    wb.save(path)


def with_normalized_gtin(record: LabelRecord) -> LabelRecord:
    return replace(record, gtin=normalize_gtin14(record.gtin))
